File: src/python/processors/ultra_fast_processor.py
#!/usr/bin/env python3
"""
Ultra-Fast Repository Processor - Optimized for Ryzen 9 3900X
Uses all 24 threads with async I/O, memory mapping, and batch processing
"""

import asyncio
import aiofiles
import multiprocessing as mp
from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor, as_completed
import os
import time
import mmap
import json
from pathlib import Path
from typing import List, Dict, Set, Optional, Generator
from dataclasses import dataclass
import hashlib
import re
from collections import defaultdict, Counter
import psutil
import numpy as np
from functools import partial
import threading
import queue
import signal

# Ultra-fast file processing with memory mapping
@dataclass
class FileChunk:
    path: str
    content: bytes
    language: str
    size: int
    hash: str

class UltraFastProcessor:
    """Ultra-optimized processor using all CPU cores and async I/O"""
    
    def __init__(self, repos_dir: str = r"\\192.168.1.66\plex3\codelupe\repos"):
        self.repos_dir = Path(repos_dir)
        
        # CPU optimization - Use ALL cores
        self.cpu_cores = psutil.cpu_count(logical=True)  # 24 threads on 3900X
        self.process_workers = self.cpu_cores - 2  # Leave 2 cores for OS
        self.io_workers = min(64, self.cpu_cores * 4)  # 4x threads for I/O
        
        print(f"🚀 Ultra-Fast Mode: {self.cpu_cores} CPU threads detected")
        print(f"🔥 Using {self.process_workers} process workers")
        print(f"⚡ Using {self.io_workers} I/O workers")
        
        # Memory optimization
        self.max_memory_gb = psutil.virtual_memory().total // (1024**3)
        self.chunk_size = min(1024 * 1024 * 10, self.max_memory_gb * 1024 * 1024 // 4)  # 10MB chunks or 1/4 RAM
        
        # Performance tracking
        self.stats = {
            'files_processed': mp.Value('i', 0),
            'bytes_processed': mp.Value('L', 0),
            'start_time': time.time(),
            'languages': defaultdict(int),
            'errors': 0
        }
        
        # File extensions for ultra-fast lookup
        self.code_extensions = {
            '.py', '.js', '.ts', '.jsx', '.tsx', '.java', '.cpp', '.c', '.h',
            '.cs', '.php', '.rb', '.go', '.rs', '.swift', '.kt', '.scala',
            '.sh', '.sql', '.r', '.m', '.pl', '.lua', '.dart', '.vim',
            '.sol', '.asm', '.hack', '.lisp', '.clj', '.hs', '.elm', '.ml'
        }
        
        # Skip patterns for ultra-fast filtering
        self.skip_dirs = {
            '.git', '.svn', '.hg', 'node_modules', '__pycache__', '.pytest_cache',
            'target', 'build', 'dist', '.gradle', 'bin', 'obj', '.idea', '.vscode',
            'vendor', 'Pods', '.pub-cache', '.nuget', 'packages', 'lib'
        }
        
        # Language mapping for fast detection
        self.ext_to_lang = {
            '.py': 'Python', '.js': 'JavaScript', '.ts': 'TypeScript',
            '.jsx': 'JavaScript', '.tsx': 'TypeScript', '.java': 'Java',
            '.cpp': 'C++', '.c': 'C', '.h': 'C/C++', '.cs': 'C#',
            '.php': 'PHP', '.rb': 'Ruby', '.go': 'Go', '.rs': 'Rust',
            '.swift': 'Swift', '.kt': 'Kotlin', '.scala': 'Scala',
            '.sh': 'Shell', '.sql': 'SQL', '.r': 'R'
        }
    
    def fast_file_scanner(self, repo_path: Path) -> Generator[Path, None, None]:
        """Ultra-fast file scanner using os.scandir()"""
        try:
            with os.scandir(repo_path) as entries:
                for entry in entries:
                    if entry.is_dir():
                        if entry.name not in self.skip_dirs:
                            yield from self.fast_file_scanner(Path(entry.path))
                    elif entry.is_file():
                        file_path = Path(entry.path)
                        if file_path.suffix.lower() in self.code_extensions:
                            # Quick size check
                            try:
                                if 100 <= entry.stat().st_size <= 1024*1024:  # 100B to 1MB
                                    yield file_path
                            except:
                                continue
        except (PermissionError, OSError):
            pass
    
    def process_file_chunk(self, file_path: Path) -> Optional[Dict]:
        """Process single file with memory mapping for speed"""
        try:
            # Fast language detection
            lang = self.ext_to_lang.get(file_path.suffix.lower(), 'Unknown')
            
            # Memory-mapped file reading for large files
            with open(file_path, 'rb') as f:
                file_size = f.seek(0, 2)
                f.seek(0)
                
                if file_size == 0 or file_size > 1024*1024:  # Skip empty or huge files
                    return None
                
                # Use memory mapping for files > 64KB
                if file_size > 65536:
                    with mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ) as mm:
                        content = mm.read()
                else:
                    content = f.read()
            
            # Fast content validation
            try:
                text_content = content.decode('utf-8', errors='ignore')
            except:
                return None
            
            if not text_content.strip():
                return None
            
            # Ultra-fast quality checks
            lines = text_content.count('\n') + 1
            if lines < 5 or lines > 2000:
                return None
            
            # Fast hash calculation
            content_hash = hashlib.blake2b(content, digest_size=16).hexdigest()
            
            # Update stats atomically
            with self.stats['files_processed'].get_lock():
                self.stats['files_processed'].value += 1
            with self.stats['bytes_processed'].get_lock():
                self.stats['bytes_processed'].value += file_size
            
            return {
                'content': text_content,
                'language': lang,
                'lines': lines,
                'size': file_size,
                'hash': content_hash,
                'path': str(file_path.relative_to(self.repos_dir))
            }
            
        except Exception:
            self.stats['errors'] += 1
            return None
    
    def process_repository_parallel(self, repo_path: Path) -> List[Dict]:
        """Process repository with parallel file processing"""
        files = list(self.fast_file_scanner(repo_path))
        if not files:
            return []
        
        results = []
        
        # Use ThreadPoolExecutor for I/O bound file processing
        with ThreadPoolExecutor(max_workers=min(32, len(files))) as executor:
            future_to_file = {
                executor.submit(self.process_file_chunk, file_path): file_path
                for file_path in files
            }
            
            for future in as_completed(future_to_file):
                try:
                    result = future.result(timeout=5)  # 5 second timeout per file
                    if result:
                        results.append(result)
                except Exception:
                    continue
        
        return results
    
    def batch_process_repositories(self, repo_paths: List[Path], batch_size: int = None) -> List[Dict]:
        """Process repositories in batches using all CPU cores"""
        if batch_size is None:
            batch_size = max(1, len(repo_paths) // self.process_workers)
        
        all_files = []
        seen_hashes = set()
        
        print(f"🔥 Processing {len(repo_paths)} repositories in batches of {batch_size}")
        
        # Process repositories in parallel using all cores
        with ProcessPoolExecutor(max_workers=self.process_workers) as executor:
            # Submit batches
            futures = []
            for i in range(0, len(repo_paths), batch_size):
                batch = repo_paths[i:i + batch_size]
                future = executor.submit(self.process_repo_batch, batch)
                futures.append(future)
            
            # Collect results with progress tracking
            completed = 0
            for future in as_completed(futures):
                try:
                    batch_results = future.result(timeout=300)  # 5 minute timeout per batch
                    
                    # Deduplicate and collect
                    for file_data in batch_results:
                        if file_data['hash'] not in seen_hashes:
                            seen_hashes.add(file_data['hash'])
                            all_files.append(file_data)
                    
                    completed += 1
                    if completed % max(1, len(futures) // 20) == 0:  # Progress every 5%
                        self.print_progress(completed, len(futures), len(all_files))
                        
                except Exception as e:
                    print(f"❌ Batch processing error: {e}")
                    continue
        
        return all_files
    
    def process_repo_batch(self, repo_batch: List[Path]) -> List[Dict]:
        """Process a batch of repositories (runs in separate process)"""
        batch_results = []
        
        for repo_path in repo_batch:
            try:
                repo_files = self.process_repository_parallel(repo_path)
                batch_results.extend(repo_files)
            except Exception:
                continue
        
        return batch_results
    
    def scan_repositories_fast(self) -> List[Path]:
        """Ultra-fast repository scanning"""
        repos = []
        
        print(f"🔍 Scanning {self.repos_dir} for repositories...")
        start_time = time.time()
        
        # Parallel directory scanning
        with ThreadPoolExecutor(max_workers=self.io_workers) as executor:
            scan_futures = []
            
            # Submit top-level directory scans
            try:
                for item in self.repos_dir.iterdir():
                    if item.is_dir() and not item.name.startswith('.'):
                        future = executor.submit(self.check_repo_validity, item)
                        scan_futures.append((future, item))
            except Exception as e:
                print(f"❌ Error scanning directory: {e}")
                return []
            
            # Collect valid repositories
            for future, repo_path in scan_futures:
                try:
                    if future.result(timeout=10):  # 10 second timeout per repo check
                        repos.append(repo_path)
                except Exception:
                    continue
        
        scan_time = time.time() - start_time
        print(f"📁 Found {len(repos)} repositories in {scan_time:.1f}s")
        
        return repos
    
    def check_repo_validity(self, repo_path: Path) -> bool:
        """Fast repository validity check"""
        try:
            # Quick check for git or code files
            has_git = (repo_path / '.git').exists()
            if has_git:
                return True
            
            # Fast code file check - sample a few files
            code_file_count = 0
            for item in repo_path.rglob('*'):
                if item.is_file() and item.suffix.lower() in self.code_extensions:
                    code_file_count += 1
                    if code_file_count >= 3:  # Found enough code files
                        return True
                if code_file_count > 50:  # Don't scan too deep
                    break
            
            return code_file_count >= 3
            
        except Exception:
            return False
    
    def print_progress(self, completed: int, total: int, files_found: int):
        """Print processing progress"""
        elapsed = time.time() - self.stats['start_time']
        rate = completed / elapsed if elapsed > 0 else 0
        
        with self.stats['files_processed'].get_lock():
            files_processed = self.stats['files_processed'].value
        with self.stats['bytes_processed'].get_lock():
            bytes_processed = self.stats['bytes_processed'].value
        
        mb_processed = bytes_processed / (1024 * 1024)
        
        print(f"🚀 Progress: {completed}/{total} batches ({completed/total*100:.1f}%) | "
              f"{files_found:,} unique files | {files_processed:,} processed | "
              f"{mb_processed:.1f}MB | {rate:.1f} batches/sec")
    
    def save_ultra_fast(self, files: List[Dict], output_file: str):
        """Ultra-fast saving with streaming JSON"""
        print(f"💾 Saving {len(files):,} files to {output_file}...")
        start_time = time.time()
        
        # Stream write for memory efficiency
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write('[\n')
            
            for i, file_data in enumerate(files):
                training_item = {
                    'text': file_data['content'],
                    'meta': {
                        'language': file_data['language'],
                        'lines': file_data['lines'],
                        'path': file_data['path'],
                        'size': file_data['size']
                    }
                }
                
                json.dump(training_item, f, ensure_ascii=False, separators=(',', ':'))
                if i < len(files) - 1:
                    f.write(',')
                f.write('\n')
                
                # Progress for large datasets
                if i % 50000 == 0 and i > 0:
                    elapsed = time.time() - start_time
                    rate = i / elapsed
                    print(f"💾 Saved {i:,}/{len(files):,} files ({rate:.0f} files/sec)")
            
            f.write(']\n')
        
        save_time = time.time() - start_time
        print(f"✅ Saved in {save_time:.1f}s ({len(files)/save_time:.0f} files/sec)")
    
    def run_ultra_fast_processing(self) -> List[Dict]:
        """Main ultra-fast processing pipeline"""
        print("🚀 ULTRA-FAST PROCESSING MODE - RYZEN 9 3900X BEAST MODE")
        print("=" * 70)
        
        # Phase 1: Ultra-fast repository scanning
        repos = self.scan_repositories_fast()
        if not repos:
            print("❌ No repositories found!")
            return []
        
        # Phase 2: Batch parallel processing
        start_time = time.time()
        files = self.batch_process_repositories(repos)
        
        # Phase 3: Final statistics
        processing_time = time.time() - start_time
        
        print(f"\n🎉 ULTRA-FAST PROCESSING COMPLETE!")
        print(f"⚡ Processing time: {processing_time:.1f}s")
        print(f"📊 Repositories processed: {len(repos):,}")
        print(f"📄 Valid files found: {len(files):,}")
        print(f"🚀 Processing rate: {len(files)/processing_time:.0f} files/sec")
        
        # Language statistics
        lang_counter = Counter(f['language'] for f in files)
        print(f"\n🔤 Top Languages:")
        for lang, count in lang_counter.most_common(10):
            print(f"   {lang}: {count:,} files")
        
        return files

def main():
    """Main function with signal handling for graceful shutdown"""
    def signal_handler(signum, frame):
        print("\n🛑 Graceful shutdown requested...")
        exit(0)
    
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    
    print("🚀 ULTRA-FAST REPOSITORY PROCESSOR")
    print("🔥 OPTIMIZED FOR RYZEN 9 3900X + 24 THREADS")
    print("=" * 70)
    
    # Check system specs
    cpu_count = psutil.cpu_count(logical=True)
    memory_gb = psutil.virtual_memory().total // (1024**3)
    print(f"💻 System: {cpu_count} CPU threads, {memory_gb}GB RAM")
    
    if cpu_count < 16:
        print("⚠️  Warning: This script is optimized for high-core CPUs (16+ threads)")
    
    # Initialize processor
    processor = UltraFastProcessor()
    
    # Run ultra-fast processing
    start_time = time.time()
    files = processor.run_ultra_fast_processing()
    
    if files:
        # Save results
        output_file = f"\\\\192.168.1.66\\plex3\\codelupe\\ultra_fast_dataset_{int(time.time())}.json"
        processor.save_ultra_fast(files, output_file)
        
        total_time = time.time() - start_time
        print(f"\n🏆 MISSION ACCOMPLISHED in {total_time:.1f}s!")
        print(f"📁 Dataset: {output_file}")
        print(f"⚡ Overall rate: {len(files)/total_time:.0f} files/sec")
    else:
        print("❌ No files processed!")

if __name__ == "__main__":
    # Optimize for multiprocessing
    mp.set_start_method('spawn', force=True)
    main()

File: src/go/processor/ultra_fast_processor.go
package main

import (
	"bufio"
	"context"
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io/fs"
	"log"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"sync/atomic"
	"time"
	"unicode/utf8"
)

func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

// FileResult represents a processed code file
type FileResult struct {
	Content  string `json:"text"`
	Language string `json:"language"`
	Lines    int    `json:"lines"`
	Size     int64  `json:"size"`
	Hash     string `json:"hash"`
	Path     string `json:"path"`
}

// TrainingData represents the final training format
type TrainingData struct {
	Text string `json:"text"`
	Meta struct {
		Language string `json:"language"`
		Lines    int    `json:"lines"`
		Path     string `json:"path"`
		Size     int64  `json:"size"`
	} `json:"meta"`
}

// ProcessorStats tracks processing statistics
type ProcessorStats struct {
	FilesProcessed int64
	BytesProcessed int64
	ReposProcessed int64
	ErrorCount     int64
	StartTime      time.Time
	LanguageCount  map[string]int64
	mu             sync.RWMutex
}

// UltraFastProcessor optimized for Ryzen 9 3900X
type UltraFastProcessor struct {
	reposDir       string
	workerCount    int
	stats          *ProcessorStats
	codeExtensions map[string]string
	skipDirs       map[string]bool
	maxFileSize    int64
	minFileSize    int64
}

// NewUltraFastProcessor creates optimized processor
func NewUltraFastProcessor(reposDir string) *UltraFastProcessor {
	// Optimize for Ryzen 9 3900X (24 threads)
	workerCount := runtime.GOMAXPROCS(0) * 4 // 96 workers for maximum throughput
	if workerCount > 128 {
		workerCount = 128 // Cap at 128 to avoid resource exhaustion
	}

	fmt.Printf("🚀 Ultra-Fast Go Processor - %d CPU cores detected\n", runtime.GOMAXPROCS(0))
	fmt.Printf("🔥 Using %d worker goroutines for MAXIMUM SPEED\n", workerCount)

	return &UltraFastProcessor{
		reposDir:    reposDir,
		workerCount: workerCount,
		stats: &ProcessorStats{
			StartTime:     time.Now(),
			LanguageCount: make(map[string]int64),
		},
		codeExtensions: map[string]string{
			".py":    "Python",
			".js":    "JavaScript",
			".ts":    "TypeScript",
			".jsx":   "JavaScript",
			".tsx":   "TypeScript",
			".java":  "Java",
			".cpp":   "C++",
			".c":     "C",
			".h":     "C/C++",
			".cs":    "C#",
			".php":   "PHP",
			".rb":    "Ruby",
			".go":    "Go",
			".rs":    "Rust",
			".swift": "Swift",
			".kt":    "Kotlin",
			".scala": "Scala",
			".sh":    "Shell",
			".sql":   "SQL",
			".r":     "R",
			".m":     "Objective-C",
			".pl":    "Perl",
			".lua":   "Lua",
			".dart":  "Dart",
			".vim":   "Vim",
			".sol":   "Solidity",
			".asm":   "Assembly",
		},
		skipDirs: map[string]bool{
			".git":          true,
			".svn":          true,
			".hg":           true,
			"node_modules":  true,
			"__pycache__":   true,
			".pytest_cache": true,
			"target":        true,
			"build":         true,
			"dist":          true,
			".gradle":       true,
			"bin":           true,
			"obj":           true,
			".idea":         true,
			".vscode":       true,
			"vendor":        true,
			"Pods":          true,
			".pub-cache":    true,
		},
		maxFileSize: 1024 * 1024, // 1MB max
		minFileSize: 100,         // 100 bytes min
	}
}

// scanRepositories scans for repositories with ultra-fast goroutines
func (p *UltraFastProcessor) scanRepositories(ctx context.Context) ([]string, error) {
	fmt.Printf("🔍 Scanning %s for repositories...\n", p.reposDir)
	start := time.Now()

	var repos []string
	var mu sync.Mutex

	// Channel for directory paths to process
	dirChan := make(chan string, p.workerCount*2)
	var wg sync.WaitGroup

	// Start scanner goroutines
	for i := 0; i < p.workerCount/4; i++ { // Use 1/4 workers for scanning
		wg.Add(1)
		go func() {
			defer wg.Done()
			for dir := range dirChan {
				if p.isValidRepository(dir) {
					mu.Lock()
					repos = append(repos, dir)
					mu.Unlock()
				}
			}
		}()
	}

	// Scan top-level directories
	entries, err := os.ReadDir(p.reposDir)
	if err != nil {
		return nil, fmt.Errorf("failed to read repos directory: %w", err)
	}

	// Send directories to workers
	go func() {
		defer close(dirChan)
		for _, entry := range entries {
			if entry.IsDir() && !strings.HasPrefix(entry.Name(), ".") {
				select {
				case dirChan <- filepath.Join(p.reposDir, entry.Name()):
				case <-ctx.Done():
					return
				}
			}
		}
	}()

	wg.Wait()

	scanTime := time.Since(start)
	fmt.Printf("📁 Found %d repositories in %.2fs\n", len(repos), scanTime.Seconds())

	return repos, nil
}

// isValidRepository checks if directory contains code files
func (p *UltraFastProcessor) isValidRepository(repoPath string) bool {
	// Quick git check
	if _, err := os.Stat(filepath.Join(repoPath, ".git")); err == nil {
		return true
	}

	// Fast code file check - sample a few files
	codeFileCount := 0
	err := filepath.WalkDir(repoPath, func(path string, d fs.DirEntry, err error) error {
		if err != nil || codeFileCount >= 5 {
			return filepath.SkipDir
		}

		if d.IsDir() {
			if p.skipDirs[d.Name()] {
				return filepath.SkipDir
			}
			return nil
		}

		ext := strings.ToLower(filepath.Ext(d.Name()))
		if _, exists := p.codeExtensions[ext]; exists {
			codeFileCount++
			if codeFileCount >= 3 {
				return filepath.SkipAll // Found enough code files
			}
		}

		return nil
	})

	return err == nil && codeFileCount >= 3
}

// processFile processes a single file with ultra-fast optimization
func (p *UltraFastProcessor) processFile(filePath string) (*FileResult, error) {
	// Fast extension check
	ext := strings.ToLower(filepath.Ext(filePath))
	language, exists := p.codeExtensions[ext]
	if !exists {
		return nil, fmt.Errorf("unsupported extension")
	}

	// Fast file size check
	info, err := os.Stat(filePath)
	if err != nil {
		return nil, err
	}

	size := info.Size()
	if size < p.minFileSize || size > p.maxFileSize {
		return nil, fmt.Errorf("file size out of range: %d", size)
	}

	// Ultra-fast file reading
	content, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	// Fast UTF-8 validation
	if !utf8.Valid(content) {
		return nil, fmt.Errorf("invalid UTF-8")
	}

	text := string(content)
	if len(strings.TrimSpace(text)) == 0 {
		return nil, fmt.Errorf("empty file")
	}

	// Fast line counting
	lines := strings.Count(text, "\n") + 1
	if lines < 5 || lines > 2000 {
		return nil, fmt.Errorf("line count out of range: %d", lines)
	}

	// Ultra-fast hash calculation
	hash := md5.Sum(content)
	hashStr := hex.EncodeToString(hash[:])

	// Update statistics atomically
	atomic.AddInt64(&p.stats.FilesProcessed, 1)
	atomic.AddInt64(&p.stats.BytesProcessed, size)

	p.stats.mu.Lock()
	p.stats.LanguageCount[language]++
	p.stats.mu.Unlock()

	// Get relative path
	relPath, err := filepath.Rel(p.reposDir, filePath)
	if err != nil {
		relPath = filePath
	}

	return &FileResult{
		Content:  text,
		Language: language,
		Lines:    lines,
		Size:     size,
		Hash:     hashStr,
		Path:     filepath.ToSlash(relPath), // Use forward slashes for JSON
	}, nil
}

// processRepository processes all files in a repository
func (p *UltraFastProcessor) processRepository(ctx context.Context, repoPath string) ([]*FileResult, error) {
	var results []*FileResult
	var mu sync.Mutex
	var wg sync.WaitGroup

	// Channel for file paths
	fileChan := make(chan string, p.workerCount)

	// Start file processing workers
	for i := 0; i < p.workerCount; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for filePath := range fileChan {
				if result, err := p.processFile(filePath); err == nil {
					mu.Lock()
					results = append(results, result)
					mu.Unlock()
				} else {
					atomic.AddInt64(&p.stats.ErrorCount, 1)
				}
			}
		}()
	}

	// Walk repository and send files to workers
	go func() {
		defer close(fileChan)
		filepath.WalkDir(repoPath, func(path string, d fs.DirEntry, err error) error {
			if err != nil {
				return nil
			}

			select {
			case <-ctx.Done():
				return filepath.SkipAll
			default:
			}

			if d.IsDir() {
				if p.skipDirs[d.Name()] {
					return filepath.SkipDir
				}
				return nil
			}

			ext := strings.ToLower(filepath.Ext(d.Name()))
			if _, exists := p.codeExtensions[ext]; exists {
				select {
				case fileChan <- path:
				case <-ctx.Done():
					return filepath.SkipAll
				}
			}

			return nil
		})
	}()

	wg.Wait()
	atomic.AddInt64(&p.stats.ReposProcessed, 1)

	return results, nil
}

// processAllRepositories processes all repositories with maximum parallelism
func (p *UltraFastProcessor) processAllRepositories(ctx context.Context) ([]*FileResult, error) {
	repos, err := p.scanRepositories(ctx)
	if err != nil {
		return nil, err
	}

	if len(repos) == 0 {
		return nil, fmt.Errorf("no repositories found")
	}

	fmt.Printf("🔥 Processing %d repositories with MAXIMUM PARALLELISM\n", len(repos))
	start := time.Now()

	var allResults []*FileResult
	var mu sync.Mutex
	var wg sync.WaitGroup

	// Channel for repositories
	repoChan := make(chan string, len(repos))

	// Start repository processing workers
	workers := p.workerCount / 8 // Use fewer workers for repo-level processing
	if workers < 1 {
		workers = 1
	}

	for i := 0; i < workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for repoPath := range repoChan {
				results, err := p.processRepository(ctx, repoPath)
				if err == nil && len(results) > 0 {
					mu.Lock()
					allResults = append(allResults, results...)
					mu.Unlock()

					fmt.Printf("✅ Processed %s: %d files\n", filepath.Base(repoPath), len(results))
				}

				// Progress update
				processed := atomic.LoadInt64(&p.stats.ReposProcessed)
				if processed%50 == 0 {
					p.printProgress(int(processed), len(repos))
				}
			}
		}()
	}

	// Send repositories to workers
	go func() {
		defer close(repoChan)
		for _, repo := range repos {
			select {
			case repoChan <- repo:
			case <-ctx.Done():
				return
			}
		}
	}()

	wg.Wait()

	processingTime := time.Since(start)
	fmt.Printf("\n🎉 ULTRA-FAST PROCESSING COMPLETE!\n")
	fmt.Printf("⚡ Processing time: %.2fs\n", processingTime.Seconds())
	fmt.Printf("📊 Repositories: %d\n", len(repos))
	fmt.Printf("📄 Files found: %d\n", len(allResults))
	fmt.Printf("🚀 Rate: %.0f files/sec\n", float64(len(allResults))/processingTime.Seconds())

	return allResults, nil
}

// printProgress prints current processing statistics
func (p *UltraFastProcessor) printProgress(completed, total int) {
	elapsed := time.Since(p.stats.StartTime)
	rate := float64(completed) / elapsed.Seconds()
	filesProcessed := atomic.LoadInt64(&p.stats.FilesProcessed)
	bytesProcessed := atomic.LoadInt64(&p.stats.BytesProcessed)

	fmt.Printf("🚀 Progress: %d/%d repos (%.1f%%) | %d files | %.1fMB | %.1f repos/sec\n",
		completed, total, float64(completed)/float64(total)*100,
		filesProcessed, float64(bytesProcessed)/(1024*1024), rate)
}

// saveDataset saves results as JSON training dataset
func (p *UltraFastProcessor) saveDataset(results []*FileResult, outputFile string) error {
	fmt.Printf("💾 Saving %d files to %s...\n", len(results), outputFile)
	start := time.Now()

	file, err := os.Create(outputFile)
	if err != nil {
		return err
	}
	defer file.Close()

	writer := bufio.NewWriter(file)
	defer writer.Flush()

	// Write JSON array start
	writer.WriteString("[\n")

	// Deduplicate by hash
	seen := make(map[string]bool)
	written := 0

	for i, result := range results {
		if seen[result.Hash] {
			continue
		}
		seen[result.Hash] = true

		// Convert to training format
		trainingData := TrainingData{
			Text: result.Content,
		}
		trainingData.Meta.Language = result.Language
		trainingData.Meta.Lines = result.Lines
		trainingData.Meta.Path = result.Path
		trainingData.Meta.Size = result.Size

		// Write JSON
		data, err := json.Marshal(trainingData)
		if err != nil {
			continue
		}

		writer.Write(data)
		if i < len(results)-1 {
			writer.WriteString(",")
		}
		writer.WriteString("\n")
		written++

		// Progress for large datasets
		if written%10000 == 0 {
			fmt.Printf("💾 Saved %d/%d files\n", written, len(results))
		}
	}

	writer.WriteString("]\n")

	saveTime := time.Since(start)
	fmt.Printf("✅ Saved %d unique files in %.2fs (%.0f files/sec)\n",
		written, saveTime.Seconds(), float64(written)/saveTime.Seconds())

	return nil
}

// printFinalStats prints comprehensive statistics
func (p *UltraFastProcessor) printFinalStats() {
	elapsed := time.Since(p.stats.StartTime)
	filesProcessed := atomic.LoadInt64(&p.stats.FilesProcessed)
	bytesProcessed := atomic.LoadInt64(&p.stats.BytesProcessed)
	reposProcessed := atomic.LoadInt64(&p.stats.ReposProcessed)
	errorCount := atomic.LoadInt64(&p.stats.ErrorCount)

	fmt.Printf("\n🏆 FINAL STATISTICS - RYZEN 9 3900X BEAST MODE\n")
	fmt.Printf("============================================================\n")
	fmt.Printf("⏱️  Total time: %.2fs\n", elapsed.Seconds())
	fmt.Printf("📁 Repositories processed: %d\n", reposProcessed)
	fmt.Printf("📄 Files processed: %d\n", filesProcessed)
	fmt.Printf("💾 Data processed: %.2fMB\n", float64(bytesProcessed)/(1024*1024))
	fmt.Printf("🚀 Processing rate: %.0f files/sec\n", float64(filesProcessed)/elapsed.Seconds())
	fmt.Printf("❌ Errors: %d\n", errorCount)

	fmt.Printf("\n🔤 Language Distribution:\n")
	p.stats.mu.RLock()
	for lang, count := range p.stats.LanguageCount {
		percentage := float64(count) / float64(filesProcessed) * 100
		fmt.Printf("   %s: %d files (%.1f%%)\n", lang, count, percentage)
	}
	p.stats.mu.RUnlock()
}

func main() {
	fmt.Printf("🚀 ULTRA-FAST GO REPOSITORY PROCESSOR\n")
	fmt.Printf("🔥 OPTIMIZED FOR RYZEN 9 3900X + 24 THREADS\n")
	fmt.Printf("============================================================\n")

	// Set GOMAXPROCS to use all cores
	runtime.GOMAXPROCS(runtime.NumCPU())
	fmt.Printf("💻 Using %d CPU cores\n", runtime.GOMAXPROCS(0))

	reposDir := getEnv("REPOS_DIR", `P:\\codelupe\\repos`)
	if len(os.Args) > 1 {
		reposDir = os.Args[1]
	}

	// Check if directory exists
	if _, err := os.Stat(reposDir); os.IsNotExist(err) {
		log.Fatalf("❌ Directory %s does not exist!", reposDir)
	}

	// Create processor
	processor := NewUltraFastProcessor(reposDir)

	// Create context with timeout
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Minute)
	defer cancel()

	// Process all repositories
	results, err := processor.processAllRepositories(ctx)
	if err != nil {
		log.Fatalf("❌ Processing failed: %v", err)
	}

	if len(results) == 0 {
		log.Fatal("❌ No files processed!")
	}

	// Save dataset
	outputFile := fmt.Sprintf("%s/ultra_fast_go_dataset_%d.json", filepath.Dir(reposDir), time.Now().Unix())
	if err := processor.saveDataset(results, outputFile); err != nil {
		log.Fatalf("❌ Failed to save dataset: %v", err)
	}

	// Print final statistics
	processor.printFinalStats()

	fmt.Printf("\n🏆 MISSION ACCOMPLISHED!\n")
	fmt.Printf("📁 Dataset: %s\n", outputFile)
}
